test/teststl.cpp

```diff
@@ -534,22 +534,49 @@ class TestStl : public TestFixture {
     }
 
     void iteratorExpression() {
-        check("std::vector<int> f();\n"
-              "std::vector<int> g();\n"
+        check("std::vector<int>& f();\n"
```

Comment:
Thanks for reviewing. The reason for using references instead of values is because doing `std::find(f().begin(), f().end(), 0)` is always incorrect when returning a vector by value. There is still no warning from cppcheck, but when we add a check for this in the future then the tests will still pass.

> maybe f() and g() return a reference to the same instance?

The checker doesn't track the references. It just sees two different expressions to get the container and concludes that there is a mismatch of iterators since writing `std::find(f().begin(), g().end(), 0)` is highly suspicious. Although unlikely, it is still possible that `f()` and `g()` return a reference to the same container. It might be better to make this a warning instead of an error.

Reply:
Maybe it should be an inconclusive error...

Are there test cases where the implementation of `f()`and `g()` is known?

Reply:
> Maybe it should be an inconclusive error...

For the warning, it would say something like, "Iterators from different expressions are used together", which is not inconclusive.

> Are there test cases where the implementation of f()and g() is known?

No there isn't. Because we only track them as variables. If we change the test case to do:

```cpp
std::vector<int>& f();
std::vector<int>& g();

void foo() {
    std::vector<int>& x = f();
    std::vector<int>& y = f();
    (void)std::find(x.begin(), y.end(), 0); // Error
}
```

This is always an error currently. Since, we already provide an error when the user assigns `f()` and `g()` to a variable, we should also provide an error when the user just calls them directly. I am not convinced this should be different(ie a warning or inconclusive).

Reply:
> For the warning, it would say something like, "Iterators from different expressions are used together", which is not inconclusive.

It is inconclusive if there is a problem or not. But well you must use the same exact same vector so it seems weird to use different expressions. So I am not sure if I like this change or not. I think we will need to use daca to verify this works well.

Reply:
> If we change the test case to do:

ok that is definitely a false positive. I do not like this error message. There could be a bug but we should not guess that.

Reply:
> There could be a bug but we should not guess that.

Yea, we probably should only give an error when they are variables that are not references or pointers.